libs/core/executors/include/hpx/executors/fork_join_executor.hpp

```diff
@@ -87,12 +89,38 @@ namespace hpx { namespace execution { namespace experimental {
                 hpx::concurrency::detail::contiguous_index_queue<std::uint32_t>;
             using queues_type =
                 std::vector<hpx::util::cache_aligned_data<queue_type>>;
-            using thread_states_type = std::vector<
-                hpx::util::cache_aligned_data<std::atomic<thread_state>>>;
-            using thread_function_helper_type = void(void*, void const*, void*,
-                std::size_t, std::size_t, loop_schedule, queues_type&,
-                thread_states_type&, hpx::lcos::local::spinlock&,
-                std::exception_ptr&);
+
+            struct region_data_type;
+            using thread_function_helper_type = void(region_data_type&,
+                std::size_t, std::size_t, queues_type&,
+                hpx::lcos::local::spinlock&, std::exception_ptr&) noexcept;
+
+            // Members that change for each parallel region.
+            struct region_data
+            {
+                // the thread state for each of the executed threads
+                std::atomic<thread_state> state_;
+
+                // The helper function that does the actual work for a single
+                // parallel region.
+                std::atomic<thread_function_helper_type*>
+                    thread_function_helper_;
+
+                // Pointers to inputs to bulk_sync_execute.
+                void* element_function_;
+                void const* shape_;
+                void* argument_pack_;
+            };
+
+            // Can't apply 'using' here as the type needs to be forward
+            // declared
+            struct region_data_type
+              : std::vector<hpx::util::cache_aligned_data<region_data>>
+            {
+                using base_type =
+                    std::vector<hpx::util::cache_aligned_data<region_data>>;
+                using base_type::base_type;
+            };
```

Comment:
I think I don't understand the need for this. Could you explain it?

Reply:
There is a circular dependency between the region_data_type and the thread function type embedded in it. This requires for the region_data_type to be forward declared, which is not possible if it's introduced with a `using` directive.